### PR TITLE
Merging Changes from GSOC23 into main

### PR DESCRIPTION
**Pull request by eltsai**
Merging the following new services into the main branch:
Low-level emulations:
1. [HTTPS and UPNP](https://github.com/honeynet/riotpot/pull/17)
2. [FTP](https://github.com/honeynet/riotpot/pull/23)

High-level services:
1. [HTTP and HTTPS](https://github.com/honeynet/riotpot/pull/25) (HTTP Updated with realistic index.html, HTTPS with self-signed cert)
2. [FTP](https://github.com/honeynet/riotpot/pull/29)
3. [Telnet](https://github.com/honeynet/riotpot/pull/31)
4. [SSH](https://github.com/honeynet/riotpot/pulls?q=is%3Apr+is%3Aclosed)

### CODE REVIEW

File: build/docker/service/custom_ftpd/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM delfer/alpine-ftp-server
+
+# Add your custom user credentials here (replace username|password with your desired values)
+ENV USERS="ftpuser|password"
```
Comment (RicYaben) [inline build/docker/service/custom_ftpd/Dockerfile]: It is a bit weird to have credentials directly in a Dockerfile. Perhaps load them from a file to docker-compose "secrets" tag and read from there.
Link: https://docs.docker.com/compose/use-secrets/

File: build/docker/docker-compose.yaml [hunk 2]
```diff
@@ -85,11 +85,54 @@ services:
...
+      - "21000-21010:21000-21010"  # Map ports for passive mode data transfer
+    environment:
+      - USERS=ftpuser|password     # Set the default FTP user (you can add more users here)
+      # Passive mode ports range
+      - MIN_PORT=21000
+      - MAX_PORT=21010
+    volumes:
+      - ./ftp_data:/ftp/ftpuser:Z       # Map the FTP data directory
+    networks:
+      default:
```
Comment (RicYaben) [inline build/docker/docker-compose.yaml]: the default networks expose the services to the Internet. This is not the behavior we are looking for with Riotpot. These services must be hidden in a virtual local network!

choose the honeypot network instead

File: build/docker/docker-compose.yaml [hunk 3]
```diff
@@ -85,11 +85,54 @@ services:
...
+    build:
+      context: ./service/custom_httpd   # Directory containing the custom Dockerfile and index.html
     container_name: http
-    expose:
-      - "80"
-    <<: *protocol
+    ports:
+      - "80:80"
+      - "443:443"   # Expose port 443 for HTTPS
+
```
Comment (RicYaben) [inline build/docker/docker-compose.yaml]: missing network

File: build/docker/docker-compose.yaml [hunk 4]
```diff
@@ -85,11 +85,54 @@ services:
...
+  telnet:
+    build:
+      context: ./service/custom_telnet
+      dockerfile: Dockerfile
+    container_name: telnet
+    restart: always
+    ports:
+      - "23:23"  # Map port 23 for Telnet
+    networks:
+      default:
```
Comment (RicYaben) [inline build/docker/docker-compose.yaml]: same as above

File: build/docker/docker-compose.yaml [hunk 5]
```diff
@@ -85,11 +85,54 @@ services:
...
+  ssh:
+    build:
+      context: ./service/custom_sshd
+      dockerfile: Dockerfile
+    container_name: ssh
+    restart: always
+    ports:
+      - "22:22"  # Map port 2222 for SSH
+    networks:
+      default:
```
Comment (RicYaben) [inline build/docker/docker-compose.yaml]: private network please

File: build/docker/docker-compose.yaml [hunk 6]
```diff
@@ -85,11 +85,54 @@ services:
...
+    build:
+      context: ./service/custom_ftpd
+      dockerfile: Dockerfile
+    container_name: ftp
+    restart: always
+    ports:
+      - "21:21"                    # Map port 21 for FTP control
+      - "21000-21010:21000-21010"  # Map ports for passive mode data transfer
+    environment:
+      - USERS=ftpuser|password     # Set the default FTP user (you can add more users here)
```
Comment (RicYaben) [inline build/docker/docker-compose.yaml]: see explanation to docker compose secrets
https://docs.docker.com/compose/use-secrets/

File: build/docker/service/custom_httpd/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+RUN sed -i '/#LoadModule ssl_module/s/^#//g' /usr/local/apache2/conf/httpd.conf \
+    && echo "Listen 443" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "<VirtualHost *:443>" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  SSLEngine on" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  SSLCertificateFile /usr/local/apache2/conf/server.crt" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  SSLCertificateKeyFile /usr/local/apache2/conf/server.key" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  DocumentRoot /usr/local/apache2/htdocs" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  ErrorLog /dev/stderr" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "  CustomLog /dev/stdout common" >> /usr/local/apache2/conf/httpd.conf \
+    && echo "</VirtualHost>" >> /usr/local/apache2/conf/httpd.conf
```
Comment (RicYaben) [inline build/docker/service/custom_httpd/Dockerfile:21]: maybe it makes more sense to append a whole document to `/usr/local/apache2/conf/httpd.conf`?

File: build/docker/service/custom_sshd/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# Install OpenSSH
+RUN apt-get update && apt-get install -y openssh-server
+
+# Install OpenSSH server
+RUN apt-get update && apt-get install -y openssh-server
+
+# Copy the SSH public key to the container's root directory
+# COPY $SSH_KEY_PUB_PATH /root/.ssh/authorized_keys
+RUN mkdir -p /root/.ssh
+#RUN echo "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPp/jbeIDrxDXW18vBuUxOj+N5msJe/elIJlgI4znfi4 riotpot@riotpot.com" >> /root/.ssh/authorized_keys
```
Comment (RicYaben) [inline build/docker/service/custom_sshd/Dockerfile]: do these comments need to stay?

File: build/docker/service/custom_sshd/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+COPY keys/riotpot_ed25519.pub /root/.ssh/authorized_keys
+
+
+# Set up SSH configuration
+RUN mkdir /var/run/sshd
+RUN echo 'root:password' | chpasswd
```
Comment (RicYaben) [inline build/docker/service/custom_sshd/Dockerfile]: same as before with the secrets. Also, I am not sure this is the behavior we need. This allows to only one set of credentials to login to the service

File: build/docker/service/custom_telnet/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+FROM ubuntu
+
+# Install Telnet
+RUN apt-get update && apt-get install -y telnet
+
+# Expose port 23 for Telnet
+EXPOSE 23
+
+# Set Telnet as the entry point
+CMD ["telnet", "localhost"]
```
Comment (RicYaben) [inline build/docker/service/custom_telnet/Dockerfile]: Is this pinging localhost? I am not sure about it

**Review by RicYaben — CHANGES_REQUESTED** [review]
In summary, we expect the following changes for this pull request:

* Place the services in the honeypot network
* Use Docker secrets or separated files as a way to add service credentials

Other than that, it looks good to me :) the rest of the thinks can be changed in future revisions

File: build/docker/service/custom_ftpd/Dockerfile (hunk 1 above)
Comment (eltsai) [reply build/docker/service/custom_ftpd/Dockerfile]: (I am replying emojis to each threads as a reminder on each task has been completed)

File: build/docker/service/custom_telnet/Dockerfile (hunk 10 above)
Comment (eltsai) [reply build/docker/service/custom_telnet/Dockerfile]: Oops I got this one wrong. The reason I got a response from port 23 is that I had a telnet server running on my local machine on port 23 (when testing I put the service on 2323). Rewriting this

**Comment by eltsai** [comment]
HTTPS:
1. ✔️Placed in the honeypot network

SSH:
1. ✔️Placed in the honeypot network
2. ✔️Used docker secret for password based authentication
3. ✔️Add instructions in README for generating `ssh_users.txt` file.

Telnet:
1. ✔️Placed in the honeypot network
2. ✔️Rewrote dockerfile

FTP: 
1. ✔️Moved username and password config into `.env` and added README.txt
2. ✔️Added a README on password and username configuration.
3. ✔️Placed in the honeypot network

**Review by RicYaben — APPROVED** [review]
Looks good :)

**Pull request merged**